src/main/java/sparksoniq/spark/udf/GroupClauseDetermineTypeUDF.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Authors: Stefan Irimescu, Can Berker Cikis
+ *
+ */
+
+package sparksoniq.spark.udf;
+
+import org.apache.spark.sql.api.java.UDF1;
+import scala.collection.mutable.WrappedArray;
+import sparksoniq.exceptions.SparksoniqRuntimeException;
+import sparksoniq.exceptions.UnexpectedTypeException;
+import sparksoniq.jsoniq.item.Item;
+import sparksoniq.jsoniq.runtime.iterator.primary.VariableReferenceIterator;
+import sparksoniq.semantics.DynamicContext;
+import sparksoniq.spark.DataFrameUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class GroupClauseDetermineTypeUDF implements UDF1<WrappedArray, List> {
```

Comment:
Why is this UDF different from the UDF used for order by?

Reply:
I think the determineTypeUDFs are more or less identical. If desired I think these two can be merged into one. In that case we will lose some precision with the error messages or need to pass the current operation type (group vs aggregate) as a parameter. We can discuss these two points further in this week's meeting